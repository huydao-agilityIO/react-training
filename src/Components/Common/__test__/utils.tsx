import React from 'react';
import * as jestFunc from '@testing-library/react';
import { ChakraProvider } from '@chakra-ui/react';

// Themes
import configThemes from '@shared/themes';

const customRender = <
  Q extends jestFunc.Queries = typeof jestFunc.queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  ui: React.ReactElement,
  options?: jestFunc.RenderOptions<Q, Container, BaseElement>
) =>
  jestFunc.render(ui, {
    ...options,
    wrapper: ({ children }: { children: React.ReactElement }) => {
      const CustomWrapper = options?.wrapper
        ? options.wrapper
        : ({ children }: { children: React.ReactElement }) => children;

      return (
        <ChakraProvider theme={configThemes}>
          <CustomWrapper>{children}</CustomWrapper>
        </ChakraProvider>
      );
    }
  });

export default { ...jestFunc, render: customRender };
