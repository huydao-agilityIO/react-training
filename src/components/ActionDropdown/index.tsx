import { memo } from 'react';
import {
  IconButton,
  List,
  ListItem,
  Popover,
  PopoverBody,
  PopoverContent,
  PopoverTrigger,
  Stack,
  useDisclosure
} from '@chakra-ui/react';

// Svg
import { EllipsisVertical } from '@shared/SVG';

interface Action {
  label: string;
  onClick?: () => void;
}

interface ActionDropdownProps {
  actions: Action[];
}

const ActionDropdown = ({ actions }: ActionDropdownProps) => {
  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <Popover isOpen={isOpen} onClose={onClose} placement="bottom-start">
      <PopoverTrigger>
        <IconButton
          aria-label="show-dropdown"
          variant="default"
          color="secondary.150"
          _hover={{
            color: 'secondary.600'
          }}
          icon={<EllipsisVertical color="secondary.150" />}
          onClick={onOpen}
        />
      </PopoverTrigger>
      <PopoverContent w={{ base: 20, md: 40 }} paddingTop={3} paddingBottom={5}>
        <PopoverBody>
          <Stack>
            <List>
              {actions?.map(({ label, onClick }) => (
                <ListItem
                  fontSize="xs"
                  fontWeight="light"
                  color="secondary.200"
                  px={5}
                  py={1}
                  _hover={{
                    cursor: 'pointer',
                    color: 'primary.300'
                  }}
                  key={label}
                  onClick={onClick}>
                  {label}
                </ListItem>
              ))}
            </List>
          </Stack>
        </PopoverBody>
      </PopoverContent>
    </Popover>
  );
};

export default memo(ActionDropdown);
