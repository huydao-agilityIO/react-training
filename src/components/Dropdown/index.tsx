import { MouseEventHandler, memo } from 'react';
import {
  Avatar,
  Button,
  Heading,
  IconButton,
  Popover,
  PopoverArrow,
  PopoverBody,
  PopoverContent,
  PopoverHeader,
  PopoverTrigger,
  Text,
  VStack
} from '@chakra-ui/react';

// Svg
import { LogoutIcon } from '@shared/SVG';

interface DropdownProps {
  fullName?: string;
  major?: string;
  isOpen?: boolean;
  onMouseEnter?: MouseEventHandler<HTMLButtonElement>;
  onMouseLeave?: MouseEventHandler<HTMLButtonElement>;
}

const Dropdown = ({
  fullName = '',
  major = 'management',
  isOpen = false,
  onMouseEnter,
  onMouseLeave
}: DropdownProps) => {
  return (
    <Popover isOpen={isOpen}>
      <PopoverTrigger>
        <IconButton
          aria-label="show-dropdown"
          variant="default"
          icon={<Avatar size={{ base: 'sm', md: 'lg' }} />}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
        />
      </PopoverTrigger>
      <PopoverContent
        p={7.5}
        w={52}
        bg="secondary.400"
        marginTop={{ sm: 0, md: 3 }}>
        <PopoverArrow bg="secondary.400" />
        <PopoverHeader>
          <VStack alignItems="flex-end" marginBottom={2}>
            <Text
              variant="secondary"
              size="xs"
              fontWeight="light"
              lineHeight="5"
              textTransform="capitalize">
              {major}
            </Text>
            <Heading
              color="light.300"
              fontSize="xs"
              fontWeight="medium"
              lineHeight="base"
              textTransform="capitalize">
              {fullName}
            </Heading>
          </VStack>
        </PopoverHeader>
        <PopoverBody>
          <VStack alignItems="flex-end">
            <Button
              variant="default"
              paddingX={0}
              color="light.300"
              rightIcon={<LogoutIcon />}>
              Log out
            </Button>
          </VStack>
        </PopoverBody>
      </PopoverContent>
    </Popover>
  );
};

export default memo(Dropdown);
