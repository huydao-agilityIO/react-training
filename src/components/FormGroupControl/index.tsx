import { ChangeEventHandler } from 'react';
import { FormControl, FormErrorMessage, Input } from '@chakra-ui/react';

interface FormGroupControlProps {
  type: string;
  id?: string;
  placeholder?: string;
  value?: string | number;
  errorMessage?: string;
  onChange?: ChangeEventHandler<HTMLInputElement>;
}

const FormGroupControl = ({
  type,
  id,
  placeholder,
  value,
  errorMessage,
  onChange
}: FormGroupControlProps) => (
  <FormControl isInvalid={!!errorMessage}>
    <Input
      type={type}
      id={id}
      placeholder={placeholder}
      value={value}
      variant={errorMessage ? 'error' : 'default'}
      onChange={onChange}
    />
    {errorMessage && <FormErrorMessage>{errorMessage}</FormErrorMessage>}
  </FormControl>
);

export default FormGroupControl;
