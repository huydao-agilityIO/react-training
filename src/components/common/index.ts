export { default as Card } from './Card';
export { default as DataTable } from './Table';
export { default as ProgressBar } from './ProgressBar';
export * from './Text';
