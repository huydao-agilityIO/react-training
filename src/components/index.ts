// Common components
export * from './common';

export * from './Author';

export { default as ActionDropdown } from './ActionDropdown';
export { default as Dropdown } from './Dropdown';
export { default as FormGroupControl } from './FormGroupControl';
export { default as Header } from './Header';
export { default as SearchBar } from './SearchBar';
export { default as Sidebar } from './Sidebar';
