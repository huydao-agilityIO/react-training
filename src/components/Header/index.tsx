import {
  HStack,
  IconButton,
  Show,
  VStack,
  useDisclosure
} from '@chakra-ui/react';

// Themes
import { breakpoints } from '@shared/themes/bases';

// Svg
import {
  MenuIcon,
  MessageIcon,
  NotificationIcon,
  SearchIcon
} from '@shared/SVG';

// Components
import { Dropdown, SearchBar } from '@shared/components';

interface HeaderProps {
  fullName?: string;
  onSearch: (value: string) => void;
}

const Header = ({ fullName, onSearch }: HeaderProps) => {
  const { md, lg } = breakpoints || {};
  const { isOpen, onOpen, onClose } = useDisclosure() || {};

  return (
    <VStack p={{ base: 5, md: 7.5 }} bg="light.300">
      <HStack w="full" justifyContent="space-between">
        <HStack>
          <Show below={`${lg}px`}>
            <IconButton
              variant="default"
              aria-label="show-sidebar"
              icon={<MenuIcon />}
            />
          </Show>
          <Show above={`${md}px`}>
            <SearchBar
              placeholder="Search here..."
              leftContent={<SearchIcon />}
              onChange={onSearch}
            />
          </Show>
        </HStack>
        <HStack spacing={{ base: 2, md: 7.5 }}>
          <NotificationIcon />
          <MessageIcon />
          <Dropdown
            fullName={fullName}
            isOpen={isOpen}
            onMouseEnter={onOpen}
            onMouseLeave={onClose}
          />
        </HStack>
      </HStack>
      <Show below={`${md}px`}>
        <SearchBar
          placeholder="Search here..."
          leftContent={<SearchIcon />}
          onChange={onSearch}
        />
      </Show>
    </VStack>
  );
};

export default Header;
