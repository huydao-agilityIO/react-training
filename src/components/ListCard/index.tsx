import { List, ListItem } from '@chakra-ui/react';

// Types
import { HospitalStaff } from '@shared/types';

// Utils
import { formatFullName } from '@shared/utils';

// Components
import { Card } from '@shared/components';

interface ListCardProps {
  cards: HospitalStaff[];
}

const ListCard = ({ cards }: ListCardProps) => (
  <List display="flex" gap={7.5}>
    {cards.map(({ id, urlAvatar, firstName, lastName, description }) => (
      <ListItem key={id}>
        <Card
          url={urlAvatar}
          description={description}
          name={formatFullName(firstName, lastName)}
        />
      </ListItem>
    ))}
  </List>
);

export default ListCard;
