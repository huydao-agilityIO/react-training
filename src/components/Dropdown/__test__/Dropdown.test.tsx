// Constants
import { VALID_INFO } from '@shared/constants';

// Components
import { Dropdown } from '@shared/components';
import { formatFullName } from '@shared/utils';

describe('Dropdown', () => {
  it('matches the snapshot', () => {
    const { container } = testLibReactUtils.render(
      <Dropdown
        fullName={formatFullName(VALID_INFO.firstName, VALID_INFO.lastName)}
        isOpen={true}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should call onOpen when the icon button is hovered', () => {
    const onOpen = jest.fn();
    const { getByRole } = testLibReactUtils.render(
      <Dropdown
        fullName={formatFullName(VALID_INFO.firstName, VALID_INFO.lastName)}
        isOpen={false}
        onOpen={onOpen}
      />
    );
    const iconButton = getByRole('button');

    testLibReactUtils.fireEvent.click(iconButton);
    expect(onOpen).toHaveBeenCalled();
  });
});
