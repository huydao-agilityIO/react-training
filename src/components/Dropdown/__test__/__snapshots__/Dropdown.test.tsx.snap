// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches the snapshot 1`] = `
<div>
  <button
    aria-controls="popover-content-:r1:"
    aria-expanded="true"
    aria-haspopup="dialog"
    aria-label="show-dropdown"
    class="chakra-button css-1p6akhc"
    id="popover-trigger-:r1:"
    type="button"
  >
    <span
      aria-hidden="true"
      class="chakra-avatar css-muc813"
      focusable="false"
    >
      <svg
        aria-label=" avatar"
        class="chakra-avatar__svg css-16ite8i"
        role="img"
        viewBox="0 0 128 128"
      >
        <path
          d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
          fill="currentColor"
        />
        <path
          d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
  <div
    class="chakra-popover__popper css-1qq679y"
    data-popper-escaped=""
    data-popper-placement="bottom"
    data-popper-reference-hidden=""
    style="visibility: visible; min-width: max-content; inset: 0 auto auto 0; --popper-transform-origin: top center; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 8px);"
  >
    <section
      aria-describedby="popover-body-:r1:"
      aria-labelledby="popover-header-:r1:"
      class="chakra-popover__content css-n8n3bl"
      id="popover-content-:r1:"
      role="dialog"
      style="transform-origin: var(--popper-transform-origin); opacity: 1; visibility: visible; transform: none;"
      tabindex="-1"
    >
      <div
        class="chakra-popover__arrow-positioner css-0"
        data-popper-arrow=""
        style="z-index: -1; --popper-arrow-size-half: calc(var(--popper-arrow-size, 8px) / 2 - 1px); --popper-arrow-offset: calc(var(--popper-arrow-size-half) * -1); position: absolute; left: 0px; transform: translate(8px, 0px);"
      >
        <div
          class="chakra-popover__arrow css-9gpjr2"
          data-popper-arrow-inner=""
          style="--popper-arrow-default-shadow: -1px -1px 0px 0 var(--popper-arrow-shadow-color); transform: rotate(45deg); top: 0px; left: 0px; width: 100%; height: 100%; position: absolute; z-index: inherit; box-shadow: var(--popper-arrow-shadow, var(--popper-arrow-default-shadow));"
        />
      </div>
      <header
        class="chakra-popover__header css-1t6b1jw"
        id="popover-header-:r1:"
      >
        <div
          class="chakra-stack css-14prjgc"
        >
          <p
            class="chakra-text css-czj0gx"
          >
            management
          </p>
          <h2
            class="chakra-heading css-2pj3d7"
          >
            Lionel Messi
          </h2>
        </div>
      </header>
      <div
        class="chakra-popover__body css-1ews2c8"
        id="popover-body-:r1:"
      >
        <div
          class="chakra-stack css-scpwb"
        >
          <button
            class="chakra-button css-1l9735s"
            type="button"
          >
            Log out
            <span
              class="chakra-button__icon css-1hzyiq5"
            >
              <svg
                aria-hidden="true"
                color="#a5adc6"
                focusable="false"
                height="12px"
                viewBox="0 0 512 512"
                width="12px"
              >
                <path
                  d="M18 32.8c-1.9.9-4.5 3.2-5.7 5.1L10 41.3v215.1c0 211.6 0 215.2 2 218.3 2 3.3 5.4 5.5 10.3 6.7 4 1 11.2-2.5 13.7-6.6 2-3.3 2-4.2 2-219V40.2l-2.7-3.3c-4.8-5.7-11.2-7.2-17.3-4.1zM188.5 179.4c-1.6.8-18.4 16.7-37.2 35.3-37.1 36.6-37 36.6-34.4 44.5 1.7 5.1 69.5 72.4 73.8 73.3 4.6.9 9.7-1 13.3-5.1 2.4-2.8 3-4.4 3-8 0-2.5-.5-5.5-1.1-6.7-.5-1.2-10.3-11.3-21.7-22.4l-20.7-20.2 163-.1c158.5 0 163.1-.1 167.2-2 10.1-4.5 11.3-18.7 2-23.7-3-1.7-12.9-1.8-167.6-2.3l-164.3-.5 20.6-19.9c11.4-11 21.4-21.2 22.2-22.7 3.1-6.1.8-14.7-5-18.4-3.5-2.3-9.5-2.8-13.1-1.1z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </section>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
